### PR TITLE
Refactor AMM, currency-adapter, router pallet to use AssetIdOf, BalanceOf

### LINKED ISSUE #644: Refactor AMM, currency-adapter, router pallet to use AssetIdOf, BalanceOf

### PR DESCRIPTION
**Pull request by ashi20192020ayush**
resolves #644

### CODE REVIEW

File: pallets/amm/src/lib.rs [hunk 1]
```diff
@@ -42,33 +40,42 @@ use frame_support::{
     transactional, Blake2_128Concat, PalletId, Twox64Concat,
 };
 use frame_system::{ensure_signed, pallet_prelude::OriginFor, RawOrigin};
-use pool_structs::PoolLiquidityAmount;
-use primitives::{Balance, CurrencyId, Rate};
+pub use pallet::*;
+use primitives::{Balance, Rate};
```
Comment (ETeissonniere) [inline pallets/amm/src/lib.rs:44]: Technically we should no longer import this

File: pallets/amm/src/lib.rs [hunk 2]
```diff
@@ -42,33 +40,42 @@ use frame_support::{
...
     use super::*;
-    use frame_support::traits::tokens::fungibles;
     use frame_system::ensure_root;
-    use primitives::CurrencyId;
+    use primitives::AssetId;
 
     #[pallet::config]
     pub trait Config<I: 'static = ()>:
-        frame_system::Config + pallet_assets::Config<AssetId = CurrencyId, Balance = Balance>
+        frame_system::Config + pallet_assets::Config<AssetId = AssetId, Balance = Balance>
```
Comment (ETeissonniere) [inline pallets/amm/src/lib.rs]: I don't think we need `AssetId` here. Did we try with `AssetIdOf` and `BalanceOf`?

File: pallets/amm/src/lib.rs [hunk 3]
```diff
@@ -42,18 +40,30 @@ use frame_support::{
...
+    <<T as Config<I>>::Assets as Inspect<<T as frame_system::Config>::AccountId>>::AssetId;
+pub type BalanceOf<T, I = ()> =
+    <<T as Config<I>>::Assets as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+
 #[frame_support::pallet]
 pub mod pallet {
     use super::*;
-    use frame_support::traits::tokens::fungibles;
     use frame_system::ensure_root;
     use primitives::CurrencyId;
```
Comment (ETeissonniere) [inline pallets/amm/src/lib.rs:68]: We no longer want to use the `primitives`

File: pallets/router/src/benchmarking.rs [hunk 4]
```diff
@@ -49,37 +68,54 @@ fn initial_set_up<T: Config<I>, I: 'static>(caller: T::AccountId) {
...
 }
 
 benchmarks_instance_pallet! {
+     where_clause {
+        where
+            BalanceOf<T, I>: FixedPointOperand,
+            AssetIdOf<T, I>: AtLeast32BitUnsigned,
+            <<T as crate::Config<I>>::Assets as Inspect<T::AccountId>>::Balance: From<u128>,
+            <<T as crate::Config<I>>::Assets as Inspect<T::AccountId>>::AssetId: From<u32>,
+            <<T as pallet_amm::Config>::Assets as Inspect<<T as frame_system::Config>::AccountId>>::Balance:
```
Comment (yz89) [inline pallets/router/src/benchmarking.rs:95]: It's too complex here. Do we have to choose to inherit pallet-amm Config on this pallet?

Comment (ashi20192020ayush) [reply pallets/router/src/benchmarking.rs:94]: @yz89 I have to add this to fix compilation error. Regarding Router Pallet implementation, you may ask @ETeissonniere or @rustaceanstayhigh

**Review by ETeissonniere — APPROVED** [review]
lgtm

**Pull request merged**